pkg/providers/cloudstack/validator.go

```diff
@@ -28,104 +30,153 @@ var restrictedUserCustomDetails = [...]string{
-	}
-	_, err := getHostnameFromUrl(datacenterConfig.Spec.ManagementApiEndpoint)
-	if err != nil {
-		return fmt.Errorf("checking management api endpoint: %v", err)
-	}
-	execConfig, err := decoder.ParseCloudStackSecret()
-	if err != nil {
-		return fmt.Errorf("parsing cloudstack secret: %v", err)
-	}
-	if execConfig.ManagementUrl != datacenterConfig.Spec.ManagementApiEndpoint {
-		return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
-			execConfig.ManagementUrl, datacenterConfig.Spec.ManagementApiEndpoint)
-	}
-
-	if err := v.validateDomainAndAccount(ctx, datacenterConfig); err != nil {
+	if err := v.generateLocalAvailabilityZones(ctx, datacenterConfig); err != nil {
 		return err
 	}
 
-	zones, errZone := v.cmk.ValidateZonesPresent(ctx, datacenterConfig.Spec.Zones)
-	if errZone != nil {
-		return fmt.Errorf("checking zones %v", errZone)
-	}
+	for _, az := range v.availabilityZones {
+		_, err := getHostnameFromUrl(az.ManagementApiEndpoint)
+		if err != nil {
+			return fmt.Errorf("checking management api endpoint: %v", err)
+		}
 
-	for _, zone := range datacenterConfig.Spec.Zones {
-		if len(zone.Network.Id) == 0 && len(zone.Network.Name) == 0 {
-			return fmt.Errorf("zone network is not set or is empty")
+		cmk, ok := v.cmks[az.CredentialsRef]
+		if !ok {
+			return fmt.Errorf("cannot find CloudStack profile for availability zone %s", az.CredentialsRef)
 		}
-		err := v.cmk.ValidateNetworkPresent(ctx, domainId, zone, zones, datacenterConfig.Spec.Account, len(zones) > 1)
+		endpoint := cmk.GetManagementApiEndpoint()
+		if endpoint != az.ManagementApiEndpoint {
+			return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
+				endpoint, az.ManagementApiEndpoint)
+		}
+
+		zoneId, err := cmk.ValidateZonePresent(ctx, az.CloudStackAvailabilityZone.Zone)
 		if err != nil {
-			return fmt.Errorf("checking network %v", err)
+			return err
+		}
+		az.CloudStackAvailabilityZone.Zone.Id = zoneId
+		if len(az.CloudStackAvailabilityZone.Zone.Network.Id) == 0 && len(az.CloudStackAvailabilityZone.Zone.Network.Name) == 0 {
+			return fmt.Errorf("zone network is not set or is empty")
+		}
+		if err := cmk.ValidateNetworkPresent(ctx, az.DomainId, az.CloudStackAvailabilityZone.Zone.Network, zoneId, az.Account, true); err != nil {
+			return err
 		}
 	}
 
 	logger.MarkPass("Datacenter validated")
 	return nil
 }
 
-func (v *Validator) validateDomainAndAccount(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if (datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account == "") ||
-		(datacenterConfig.Spec.Domain == "" && datacenterConfig.Spec.Account != "") {
-		return fmt.Errorf("both domain and account must be specified or none of them must be specified")
+func (v *Validator) generateLocalAvailabilityZones(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
```

Comment:
it seems like we're doing two things in this function - generating local AZ's, and also validating their accounts/domains. I'd suggest simplifying this to just generate the AZ's, and do the validation of them in ValidateCloudStackDatacenterConfig